src/lib/flow/sol-flow-node-options.c

```diff
@@ -478,104 +476,176 @@ direction_vector_parse(const struct sol_flow_node_options_member_description *me
     }
 
-    return NULL;
+    m->byte = v;
+    return 0;
 }
 
 static int
-split_option(const char *input, const char **key, unsigned int *key_len, const char **value)
+string_parse(const char *value, struct sol_flow_node_named_options_member *m)
 {
-    const char *equal = strchr(input, '=');
+    char *s;
 
-    if (!equal || equal == input || equal + 1 == '\0')
-        return -EINVAL;
+    s = strdup(value);
+    if (!s)
+        return -errno;
 
-    *key = input;
-    *key_len = equal - input;
-    *value = equal + 1;
+    m->string = s;
     return 0;
 }
 
+static int(*const options_parse_functions[]) (const char *, struct sol_flow_node_named_options_member *) = {
+    NULL,
+    boolean_parse,
+    byte_parse,
+    irange_parse,
+    drange_parse,
+    rgb_parse,
+    direction_vector_parse,
+    string_parse,
+};
+
+SOL_API int
+sol_flow_node_named_options_parse_member(
+    struct sol_flow_node_named_options_member *m,
+    const char *value)
+{
+    int r;
+
+    if (m->type == 0) {
+        r = -EINVAL;
+        SOL_DBG("Unitialized member type for name=\"%s\": \"%s\"", m->name, value);
+    } else if (m->type < ARRAY_SIZE(options_parse_functions)) {
+        r = options_parse_functions[m->type](value, m);
+        if (r < 0) {
+            SOL_DBG("Invalid value '%s' for option name='%s' of type='%s'",
+                value, m->name, sol_flow_node_options_member_type_to_string(m->type));
+        }
+    } else {
+        r = -ENOTSUP;
+        SOL_DBG("Invalid option member type for name=\"%s\": \"%s\"", m->name, value);
+    }
+
+    return r;
+}
+
+static void
+set_member(
+    const struct sol_flow_node_options_member_description *mdesc,
+    enum sol_flow_node_options_member_type mdesc_type,
+    const struct sol_flow_node_named_options_member *m,
+    struct sol_flow_node_options *opts)
+{
+    char **s;
+    void *mem = get_member_memory(mdesc, opts);
+
+    switch (mdesc_type) {
+    case SOL_FLOW_NODE_OPTIONS_MEMBER_BOOLEAN:
+        *(bool *)mem = m->boolean;
+        break;
+    case SOL_FLOW_NODE_OPTIONS_MEMBER_BYTE:
+        *(unsigned char *)mem = m->byte;
+        break;
+    case SOL_FLOW_NODE_OPTIONS_MEMBER_IRANGE:
+        *(struct sol_irange *)mem = m->irange;
+        break;
+    case SOL_FLOW_NODE_OPTIONS_MEMBER_DRANGE:
+        *(struct sol_drange *)mem = m->drange;
+        break;
+    case SOL_FLOW_NODE_OPTIONS_MEMBER_RGB:
+        *(struct sol_rgb *)mem = m->rgb;
+        break;
+    case SOL_FLOW_NODE_OPTIONS_MEMBER_DIRECTION_VECTOR:
+        *(struct sol_direction_vector *)mem = m->direction_vector;
+        break;
+    case SOL_FLOW_NODE_OPTIONS_MEMBER_STRING:
+        s = mem;
+        free(*s);
+        *s = strdup(m->string);
+    default:
+        /* Not reached. */
+        /* TODO: Create ASSERT_NOT_REACHED() macro.*/
+        break;
+    }
+}
+
+static const struct sol_str_table member_str_to_type[] = {
+    SOL_STR_TABLE_ITEM("boolean", SOL_FLOW_NODE_OPTIONS_MEMBER_BOOLEAN),
```

Comment:
Maybe you can join this table with `options_parse_functions[]`? By making it a `string` -> `function ptr` mapping, and indexing it with the `enum sol_flow_node_options_member_type`, only one table will need to be maintained.

The `sol_str_table` will need to be changed to have a function that also gives the index.